src/llmq/signing_shares.cpp

```diff
@@ -1591,60 +1611,106 @@ void CSigSharesManager::BanNode(NodeId nodeId)
-
     while (!workInterrupt) {
         RemoveBannedNodeStates();
+        SendMessages();
+        Cleanup();
 
-        bool fMoreWork = ProcessPendingSigShares();
-        SignPendingSigShares();
+        workInterrupt.sleep_for(std::chrono::milliseconds(100));
+    }
+}
 
-        if (TicksSinceEpoch<std::chrono::milliseconds>(SystemClock::now()) - lastSendTime > 100) {
-            SendMessages();
-            lastSendTime = TicksSinceEpoch<std::chrono::milliseconds>(SystemClock::now());
-        }
+void CSigSharesManager::WorkDispatcherThreadMain()
+{
+    while (!workInterrupt) {
+        // Dispatch all pending signs (individual tasks)
+        DispatchPendingSigns();
 
-        Cleanup();
+        // If there's processing work, spawn a helper worker
+        DispatchPendingProcessing();
 
-        // TODO Wakeup when pending signing is needed?
-        if (!fMoreWork && !workInterrupt.sleep_for(std::chrono::milliseconds(100))) {
-            return;
-        }
+        // Always sleep briefly between checks
+        workInterrupt.sleep_for(std::chrono::milliseconds(10));
     }
 }
 
-void CSigSharesManager::AsyncSign(CQuorumCPtr quorum, const uint256& id, const uint256& msgHash)
+void CSigSharesManager::DispatchPendingSigns()
 {
-    LOCK(cs_pendingSigns);
-    pendingSigns.emplace_back(std::move(quorum), id, msgHash);
+    // Swap out entire vector to avoid lock thrashing
+    std::vector<PendingSignatureData> signs;
+    {
+        LOCK(cs_pendingSigns);
+        signs.swap(pendingSigns);
+    }
+
+    // Dispatch all signs to worker pool
+    for (auto& work : signs) {
+        if (workInterrupt) break;
+
+        workerPool.push([this, work = std::move(work)](int) {
+            SignAndProcessSingleShare(std::move(work));
+        });
+    }
 }
 
-void CSigSharesManager::SignPendingSigShares()
+void CSigSharesManager::DispatchPendingProcessing()
 {
-    std::vector<PendingSignatureData> v;
-    WITH_LOCK(cs_pendingSigns, v.swap(pendingSigns));
-
-    for (const auto& [pQuorum, id, msgHash] : v) {
-        auto opt_sigShare = CreateSigShare(*pQuorum, id, msgHash);
-
-        if (opt_sigShare.has_value() && opt_sigShare->sigShare.Get().IsValid()) {
-            auto sigShare = *opt_sigShare;
-            ProcessSigShare(sigShare, pQuorum);
-
-            if (IsAllMembersConnectedEnabled(pQuorum->params.type, m_sporkman)) {
-                LOCK(cs);
-                auto& session = signedSessions[sigShare.GetSignHash()];
-                session.sigShare = sigShare;
-                session.quorum = pQuorum;
-                session.nextAttemptTime = 0;
-                session.attempt = 0;
-            }
+    // Check if there's work, spawn a helper if so
+    bool hasWork = false;
+    {
+        LOCK(cs);
+        hasWork = std::any_of(nodeStates.begin(), nodeStates.end(),
+            [](const auto& entry) {
+                return !entry.second.pendingIncomingSigShares.Empty();
+            });
+    }
+
+    if (hasWork) {
```

Comment:
Shouldn't queue when there is no worker left to help, that's just noop queue entries added every 10ms which will spawn short living workers once the job is done.
```suggestion
    if (hasWork && workerPool.n_idle() > 0) {
```